include/BingoCpp/gradient_mixin.h

```diff
@@ -0,0 +1,61 @@
+#ifndef BINGOCPP_INCLUDE_BINGOCPP_GRADIENT_MIXIN_H_
+#define BINGOCPP_INCLUDE_BINGOCPP_GRADIENT_MIXIN_H_
+
+#include <Eigen/Dense>
+#include <bingocpp/equation.h>
+#include <bingocpp/fitness_function.h>
+
+#include <functional>
+#include <cmath>
+
+namespace bingo {
+
+class GradientMixin {
+ public:
+  virtual std::tuple<double, Eigen::VectorXd> GetIndividualFitnessAndGradient(Equation &individual) const = 0;
+};
+
+class VectorGradientMixin : public GradientMixin {
+ public:
+  VectorGradientMixin(TrainingData *training_data = nullptr, std::string metric = "mae");
+
+  std::tuple<double, Eigen::VectorXd> GetIndividualFitnessAndGradient(Equation &individual) const;
+
+  virtual std::tuple<Eigen::VectorXd, Eigen::ArrayXXd> GetFitnessVectorAndJacobian(Equation &individual) const = 0;
+
+ protected:
+  static double mean_absolute_error(const Eigen::VectorXd &fitness_vector) {
+    return fitness_vector.array().abs().mean();
+  }
+
+  static double mean_squared_error(const Eigen::VectorXd &fitness_vector) {
+    return fitness_vector.array().square().mean();
+  }
+
+  static double root_mean_squared_error(const Eigen::VectorXd &fitness_vector) {
+    return sqrt(fitness_vector.array().square().mean());
+  }
+
+  static Eigen::VectorXd mean_absolute_error_derivative(
+      const Eigen::VectorXd &fitness_vector, const Eigen::ArrayXXd &fitness_partials) {
+    return (fitness_partials.rowwise() * fitness_vector.transpose().array().sign()).rowwise().mean();
+  }
+
+  static Eigen::VectorXd mean_squared_error_derivative(
+      const Eigen::VectorXd &fitness_vector, const Eigen::ArrayXXd &fitness_partials) {
+    return 2.0 * (fitness_partials.rowwise() * fitness_vector.transpose().array()).rowwise().mean();
+  }
+
+  static Eigen::VectorXd root_mean_squared_error_derivative(
+      const Eigen::VectorXd &fitness_vector, const Eigen::ArrayXXd &fitness_partials) {
+    return 1.0/sqrt(fitness_vector.array().square().mean()) *
+    (fitness_partials.rowwise() * fitness_vector.transpose().array()).rowwise().mean();
+  }
+
+ private:
+  std::function<double(Eigen::ArrayXXd)> metric_function_;
+  std::function<Eigen::ArrayXXd(Eigen::ArrayXXd, Eigen::ArrayXXd)> metric_derivative_;
```

Comment:
I was wondering if std::function wrappers or pure function pointers are preferred here. In [`fitness_function.h`](https://github.com/nasa/bingocpp/blob/develop/include/bingocpp/fitness_function.h#L74) we use pure function pointers, but this requires [const casting when using them.](https://github.com/nasa/bingocpp/blob/develop/include/bingocpp/fitness_function.h#L85) I don't really have a preference for either method, so I wanted to know what you think.

Reply:
std::function wrappers work for me

Could you update fitness_function.h to match?

Reply:
sure thing, I will change fitness function to match too!